ls1intum/Artemis
# `Development`: Restructure docker compose setups

### Checklist
#### General


- [x] I tested these configs locally (Artemis) and with multiple docker setups (Artemis, MySQL, Jenkins, GitLab, Atlassian Stack).

### Motivation and Context



So far docker compose files were mainly written as little helpers for development but not for the usage in production, development, and the combination of multiple docker compose services.

This PR structures the docker compose files and documents the setups and services.

The Artemis Dockerfile was also refactored with the goal to build the whole application without the need of a .war file beforehand and therefore documenting the build process.

### Description


This PR lead to the idea of restructuring the configuration files in the following PR, but is NOT depended on it:

- #5944

### Migration Path

#### Bamboo Tests
I re-added and updated the E2E bamboo scripts which @TheZoker and I debugged manually last week to get the E2E tests stable again. So these scripts have to be activated again (load from version control instead of inline) after merging this PR.

Links to the bamboo plans:
https://bamboo.ase.in.tum.de/build/admin/edit/editBuildTasks.action?buildKey=ARTEMIS-AETG-DA
https://bamboo.ase.in.tum.de/build/admin/edit/editBuildTasks.action?buildKey=ARTEMIS-AECF-DA

The tasks are 2x ``CLEANUP`` (has to point to the corresponding ``./.bamboo/.../cleanup.sh``) and ``Deploy Artemis and run Cypress tests`` (has to point to the corresponding ``./.bamboo/.../execute.sh``) for each plan. After switching the script location from ``Inline`` to ``File`` the corresponding paths are probably already set.

#### Docker volumes
As some docker volumes were renamed, you might have to rename these if you want to keep your state of your docker containers.

Do so by looking up the current volume names:

```bash
docker volume ls
```

The old volume names might have prefixes like ``docker_``.
This shows a table with the ``old_volume`` name and the ``new_volume`` name:

old_volume | new_volume
--- | ---
`[PREFIX]artemis-bamboo` | `artemis-bamboo-data`
`[PREFIX]artemis-bitbucket` | `artemis-bitbucket-data`
`[PREFIX]artemis-jira` | `artemis-jira-data`
`[PREFIX]artemis-gitlab-data` | `artemis-gitlab-data`
`[PREFIX]artemis-gitlab-logs` | `artemis-gitlab-logs`
`[PREFIX]artemis-gitlab-config` | `artemis-gitlab-config`
`[PREFIX]artemis-jenkins-data` | `artemis-jenkins-data`
`[PREFIX]artemis-mysql-data` | `artemis-mysql-data`
`[PREFIX]artemis-gitlabci-runner-config` | `artemis-gitlabci-runner-config`

Then the following procedure should be done for each volume as there is no other way to rename a volume apparently in docker:
```bash
docker volume create --name <new_volume>
docker run --rm -it -v <old_volume>:/from -v <new_volume>:/to alpine ash -c "cd /from ; cp -av . /to"
docker volume rm <old_volume>
```

#### Hard-coded Docker IPs
Remove hard-coded docker IP addresses from your ``application-local.yml``.

### Steps for Testing

- Check/Read the docs concerning the [Dockerfile](https://artemis-platform--5915.org.readthedocs.build/en/5915/dev/setup/#run-the-server-via-docker) and the [Docker Compose setups](https://artemis-platform--5915.org.readthedocs.build/en/5915/dev/setup/#alternative-docker-compose-setup)
- Test each or a subset of the following setups:
  - **Artemis-Prod-MySQL** (docker-compose.yml in the docker directory)
  - **Atlassian** (separate Atlassian stack with build agent, check that builds complete and start in a separate docker container on the docker host)
  - **GitLab-Jenkins** (separate GitLab-Jenkins stack)
  - **MySQL** (separate MySQL container)
  - **Prometheus-Grafana** (separate Prometheus Grafana stack)
- Test the Bamboo E2E Setup (can be tested in this [customized Bamboo Build pipeline](https://bamboo.ase.in.tum.de/chain/viewChain.action?planKey=AMPU-AETM81))

Note: If you do not have time to build the Docker Image by yourself, you can also pull it from:
``ghcr.io/ls1intum/artemis:pr-5915``

### Review Progress






#### Code Review
#### Manual Tests of these setups:
- [x] Artemis-Prod-MySQL 1
- [x] Artemis-Prod-MySQL 2
- [x] Atlassian 1
- [ ] Atlassian 2
- [x] GitLab-Jenkins 1
- [ ] GitLab-Jenkins 2
- [x] MySQL 1
- [x] MySQL 2
- [x] Prometheus-Grafana 1
- [ ] Prometheus-Grafana 2
- [x] Bamboo E2E Setup 1
- [ ] Bamboo E2E Setup 2
#### Exam Mode Test
